src/RevProc.cc

```diff
@@ -2369,94 +2424,67 @@ void RevProc::InitEcallTable(){
-    LSQueue->insert({make_lsq_hash(req.DestReg, req.RegType, req.Hart), req});
-    mem->ReadVal(HartToExec,
-                 straddr + ECALL.string.size(),
-                 ECALL.buf.data(),
-                 req,
-                 REVMEM_FLAGS(0));
-    ECALL.bytesRead = 1;
-    DependencySet(HartToExec, 10, false);
-    rtval = ECALL_status_t::CONTINUE;
-  }
-  return rtval;
-}
-
-/*
- * This is the function that is called when an ECALL exception is detected inside ClockTick
- * - Currently the only way to set this exception is by Ext->Execute(....) an ECALL instruction
- *
- * Eventually this will be integrated into a TrapHandler however since ECALLs are the only
- * supported exceptions at this point there is no need just yet.
- */
+// This is the function that is called when an ECALL exception is detected inside ClockTick
+// Currently the only way to set this exception is by Ext->Execute(....) an ECALL instruction
+// Eventually this will be integrated into a TrapHandler however since ECALLs are the only
+// supported exceptions at this point there is no need just yet.
 void RevProc::ExecEcall(RevInst& inst){
   // a7 register = ecall code
-  auto EcallCode = RegFile->GetX<uint64_t>(RevReg::a7);
+  auto EcallCode = RegFile->GetX<uint64_t>(17);
   auto it = Ecalls.find(EcallCode);
   if( it != Ecalls.end() ){
     // call the function
-    ECALL_status_t status = it->second(this, inst);
+    // TODO: Maybe change this with the new Hart infrastructure? 
+    EcallStatus status = it->second(this, inst);
 
     // Trap handled... 0 cause registers
     RegFile->RV64_SCAUSE = uint64_t(status);
     RegFile->RV32_SCAUSE = uint32_t(status);
 
     // For now, rewind the PC and keep executing the ECALL until we
     // have completed
-    if(ECALL_status_t::SUCCESS != status){
+    if(EcallStatus::SUCCESS != status){
       RegFile->AdvancePC( -int32_t(inst.instSize) );
     }
   } else {
     output->fatal(CALL_INFO, -1, "Ecall Code = %" PRIu64 " not found", EcallCode);
   }
 }
 
-uint32_t RevProc::GetActiveThreadID(){
-  uint32_t ActiveThreadID = 0;
-  if( HartToDecode != _REV_INVALID_HART_ID_ ){
-    if( AssignedThreads.size() >= HartToDecode ){
-      ActiveThreadID = AssignedThreads.at(HartToDecode)->GetThreadID();
-    }
-    else {
-      output->fatal(CALL_INFO, 1, "HartToDecode = %" PRIu32 " but there are only %" PRIu64 " threads assigned to this Proc. This might be a bug\n", HartToDecode, AssignedThreads.size());
+// TODO: (BEFORE MERGE) Potentially add a faster way to find available Harts
+// Also I don't think we ever get here without first knowing that there is a Hart available
+// so maybe we can just assign the thread to the first available hart and be done with it
+void RevProc::AssignThread(const uint32_t& ThreadID){
+  bool ThreadAssigned = false;
+  // Check for available Hart 
+  for( auto Hart : Harts ){
```

Comment:
For performance, it should be `auto& Hart : Harts`.

As it is written now, it causes the atomic reference count of each `std::shared_ptr` element to be incremented and decremented at each loop iteration.

Reply:
Fixed!